man/RNAsubopt.1
2.1.7

.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH RNASUBOPT "1" "April 2014" "RNAsubopt 2.1.7" "User Commands"
.SH NAME
RNAsubopt \- manual page for RNAsubopt 2.1.7
.SH SYNOPSIS
.B RNAsubopt
[\fIOPTIONS\fR]...
.SH DESCRIPTION
RNAsubopt 2.1.7
.PP
calculate suboptimal secondary structures of RNAs
.PP
reads RNA sequences from stdin and (in the default \fB\-e\fR mode) calculates all
suboptimal secondary structures within a user defined energy range above the
minimum free energy (mfe). It prints the suboptimal structures in dot\-bracket
notation followed by the energy in kcal/mol to stdout. Be careful, the number
of structures returned grows exponentially with both sequence length and energy
range.
.PP
Alternatively, when used with the \fB\-p\fR option, RNAsubopt produces Boltzmann
weighted samples of secondary structures.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-detailed\-help\fR
Print help, including all details and hidden
options, and exit
.TP
\fB\-\-full\-help\fR
Print help, including hidden options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.SS "General Options:"
.IP
Below are command line options which alter the general behavior of this
program
.TP
\fB\-C\fR, \fB\-\-constraint\fR
Calculate structures subject to constraints.
(default=off)
.IP
The program reads first the sequence, then a string containing constraints on
the structure encoded with the symbols:
.IP
\&. (no constraint for this base)
.IP
| (the corresponding base has to be paired
.IP
x (the base is unpaired)
.IP
< (base i is paired with a base j>i)
.IP
\f(CW> (base i is paired with a base j<i)\fR
.IP
and matching brackets ( ) (base i pairs base j)
.IP
With the exception of "|", constraints will disallow all pairs conflicting
with the constraint. This is usually sufficient to enforce the constraint,
but occasionally a base may stay unpaired in spite of constraints. PF folding
ignores constraints of type "|".
.TP
\fB\-\-canonicalBPonly\fR
Remove non\-canonical base pairs from the
structure constraint
.IP
(default=off)
.TP
\fB\-\-noconv\fR
Do not automatically substitude nucleotide
"T" with "U"
.IP
(default=off)
.SS "Algorithms:"
.IP
Select the algorithms which should be applied to the given RNA sequence.
.TP
\fB\-e\fR, \fB\-\-deltaEnergy\fR=\fIrange\fR
Compute suboptimal structures with energy in a
certain range of the optimum (kcal/mol).
Default is calculation of mfe structure only.
.TP
\fB\-\-deltaEnergyPost\fR=\fIrange\fR
Only print structures with energy within range
of the mfe after post reevaluation of
energies.
.IP
Useful in conjunction with \fB\-logML\fR, \fB\-d1\fR or \fB\-d3\fR: while the \fB\-e\fR option specifies
the range before energies are re\-evaluated, this option specifies the maximum
energy after re\-evaluation.
.TP
\fB\-s\fR, \fB\-\-sorted\fR
Sort the suboptimal structures by energy.
(default=off)
.IP
Since the sort in is done in memory, this becomes impractical when the number
of structures produced goes into millions. In such cases better pipe the
output through "sort +1n".
.TP
\fB\-p\fR, \fB\-\-stochBT\fR=\fInumber\fR
Instead of producing all suboptimals in an
energy range, produce a random sample of
suboptimal structures, drawn with
probabilities equal to their Boltzmann
weights via stochastic backtracking in the
partition function. The \fB\-e\fR and \fB\-p\fR options are
mutually exclusive.
.TP
\fB\-S\fR, \fB\-\-pfScale\fR=\fIscaling\fR factor
In the calculation of the pf use scale*mfe as
an estimate for the ensemble free energy
(used to avoid overflows). Needed by
stochastic backtracking
.IP
The default is 1.07, useful values are 1.0 to 1.2. Occasionally needed for
long sequences.
You can also recompile the program to use double precision (see the README
file).
.TP
\fB\-c\fR, \fB\-\-circ\fR
Assume a circular (instead of linear) RNA
molecule.
.IP
(default=off)
.TP
\fB\-D\fR, \fB\-\-dos\fR
Compute density of states instead of secondary
structures
.IP
(default=off)
.IP
This option enables the evaluation of the number of secondary structures in
certain energy bands arround the MFE.
.TP
\fB\-z\fR, \fB\-\-zuker\fR
Compute Zuker suboptimals instead of all
suboptimal structures within an engery band
arround the MFE.
.IP
(default=off)
.TP
\fB\-g\fR, \fB\-\-gquad\fR
Incoorporate G\-Quadruplex formation into the
structure prediction algorithm
(no support of G\-quadruplex prediction for
stochastic backtracking and Zuker\-style
suboptimals yet)
.IP
(default=off)
.SS "Model Details:"
.TP
\fB\-T\fR, \fB\-\-temp\fR=\fIDOUBLE\fR
Rescale energy parameters to a temperature of
temp C. Default is 37C.
.TP
\fB\-4\fR, \fB\-\-noTetra\fR
Do not include special tabulated stabilizing
energies for tri\-, tetra\- and hexaloop
hairpins. Mostly for testing.
.IP
(default=off)
.TP
\fB\-d\fR, \fB\-\-dangles\fR=\fIINT\fR
How to treat "dangling end" energies for
bases adjacent to helices in free ends and
multi\-loops
.IP
(default=`2')
.IP
With \fB\-d1\fR only unpaired bases can participate in at most one dangling end,
this is the default for mfe folding but unsupported for the partition
function folding.
.IP
With \fB\-d2\fR this check is ignored, dangling energies will be added for the bases
adjacent to a helix on both sides in any case; this is the default for
partition function folding (\fB\-p\fR).
The option \fB\-d0\fR ignores dangling ends altogether (mostly for debugging).
With \fB\-d3\fR mfe folding will allow coaxial stacking of adjacent helices in
multi\-loops. At the moment the implementation will not allow coaxial stacking
of the two interior pairs in a loop of degree 3 and works only for mfe
folding.
.IP
Note that by default (as well as with \fB\-d1\fR and \fB\-d3\fR) pf and mfe folding treat
dangling ends differently. Use \fB\-d2\fR in addition to \fB\-p\fR to ensure that both
algorithms use the same energy model.
.TP
\fB\-\-noLP\fR
Produce structures without lonely pairs
(helices of length 1).
.IP
(default=off)
.IP
For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.
.TP
\fB\-\-noGU\fR
Do not allow GU pairs
.IP
(default=off)
.TP
\fB\-\-noClosingGU\fR
Do not allow GU pairs at the end of helices
.IP
(default=off)
.TP
\fB\-\-logML\fR
Recalculate energies of structures using a
logarithmic energy function for multi\-loops
before output.  (default=off)
.IP
This option does not effect structure generation, only the energies that are
printed out. Since logML lowers energies somewhat, some structures may be
missing.
.TP
\fB\-\-betaScale\fR=\fIDOUBLE\fR
Set the scaling of the Boltzmann factors
(default=`1.')
.IP
The argument provided with this option enables to scale the thermodynamic
temperature used in the Boltzmann factors independently from the temperature
used to scale the individual energy contributions of the loop types. The
Boltzmann factors then become exp(\fB\-dG\fR/(kT*betaScale)) where k is the
Boltzmann constant, dG the free energy contribution of the state and T the
absolute temperature.
.TP
\fB\-P\fR, \fB\-\-paramFile\fR=\fIparamfile\fR
Read energy parameters from paramfile, instead
of using the default parameter set.
.IP
A sample parameter file should accompany your distribution.
See the RNAlib documentation for details on the file format.
.TP
\fB\-\-nsp\fR=\fISTRING\fR
Allow other pairs in addition to the usual
AU,GC,and GU pairs.
.IP
Its argument is a comma separated list of additionally allowed pairs. If the
first character is a "\-" then AB will imply that AB and BA are allowed
pairs.
e.g. RNAfold \fB\-nsp\fR \fB\-GA\fR  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.
.SH AUTHOR

Ivo L Hofacker, Stefan Wuchty, Walter Fontana, Ronny Lorenz
.SH REFERENCES
.I If you use this program in your work you might want to cite:

R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
"ViennaRNA Package 2.0",
Algorithms for Molecular Biology: 6:26 

I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
"Fast Folding and Comparison of RNA Secondary Structures",
Monatshefte f. Chemie: 125, pp 167-188


S. Wuchty, W. Fontana, I. L. Hofacker and P. Schuster (1999),
"Complete Suboptimal Folding of RNA and the Stability of Secondary Structures",
Biopolymers: 49, pp 145-165

.I The energy parameters are taken from:

D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
"Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure",
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292

D.H Turner, D.H. Mathews (2009),
"NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure",
Nucleic Acids Research: 38, pp 280-282
.SH "REPORTING BUGS"
If in doubt our program is right, nature is at fault.
.br
Comments should be sent to rna@tbi.univie.ac.at.


NEWS
Some Changes relative to ViennaRNA-1.03

New energy parameters used in the folding routines.
Prediction of minimum free energy structures now includes contributions
from terminal unpaired nucleotides (so called dangling ends).
The partition function treats these contributions approximately.

Energies for GU pairs in are now taken from
L. He et al., Biochemistry 1991, 30, p 11124

Bulges of length 1 now get an additional stacking energy.

The folding routines accept  a list of nonstandard basepairs that should be
allowed to occur. Also, when doing constrained folding, forced pairs will
be honoured even if the pair is usally not allowed (not AU, GC, or GU).
RNAeval should be useful now for structures containing nonstandard pairs.

Some support for constrained folding using partition function: A constraint
of the form i-j must pair or i must not pair will be honoured by disallowing 
all pairs conflicting with the constraint.

Both folding algorthms use less memory and should be a bit faster for long
sequences.

Several smaller bug fixes.


Subopt/RNAsubopt.c
/* Last changed Time-stamp: <1999-02-06 21:56:11 ivo> */
/*                
		Ineractive Access to suboptimal folding

			   c Ivo L Hofacker
			  Vienna RNA package
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <string.h>
#include "fold.h"
#include "fold_vars.h"
#include "utils.h"
static char rcsid[] = "$Id: RNAsubopt.c,v 1.2 1998/05/19 16:31:17 ivo rel ivo $";

#define PRIVATE static

static char  scale[] = "....,....1....,....2....,....3....,....4"
                       "....,....5....,....6....,....7....,....8";

PRIVATE void usage(void);

int sorted=0;
int delta=100;
char *sequence;

int LODOS_ONLY = 0;

extern void  read_parameter_file(const char fname[]);
extern void subopt(int delta);

/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
   char  *line;
   char *structure = NULL, *cstruc = NULL;
   char  fname[21];
   char  ParamFile[256] = "";
   char  ns_bases[33] = "", *c;
   int   i, length, l, sym, r;
   float energy, min_en;
   int   istty;
   float deltaf;
   
   do_backtrack = 1;
   dangles = 2;
   sequence=NULL;
   for (i=1; i<argc; i++) {
      if (argv[i][0]=='-') 
	switch ( argv[i][1] )
	  {
	  case 'T':  if (argv[i][2]!='\0') usage();
	    if(i==argc-1) usage();
	    r=sscanf(argv[++i], "%f", &temperature);
	    if (!r) usage();
	    break;
	  case 'n':
	    if ( strcmp(argv[i], "-noGU" )==0) noGU=1;
	    if ( strcmp(argv[i], "-noCloseGU" ) ==0) no_closingGU=1;
	    if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	    if ( strcmp(argv[i], "-nsp") ==0) {
	      if (i==argc-1) usage();
	      r=sscanf(argv[++i], "%32s", ns_bases);
	      if (!r) usage();
	    }
	    break;
	  case '4':
	    tetra_loop=0;
	    break;
 	  case 'C': 
 	    fold_constrained=1; 
 	    break; 
	  case 'd': dangles=0;
	    if (strcmp(argv[i],"-d2")==0) dangles=2;
	    /* danlges == 1 not allowed in subopt() */
	    if (strcmp(argv[i],"-d1")==0) usage();
	    break;
	  case 'P':
	    if (i==argc-1) usage();
	    r=sscanf(argv[++i], "%255s", ParamFile);
	    if (!r) usage();
	    break;
	  case 's':
	    sorted=1;
	    break;
	  case 'l':
	    if (strcmp(argv[i],"-lodos")==0) {
	      LODOS_ONLY=1;
	      sorted = 1;
	    }
	    else {
	      if (strcmp(argv[i],"-logML")==0) {
		logML=1;
		break;
	      }
	      else usage();
	    }
	    break;
	  case 'e':
	    r=sscanf(argv[++i], "%f", &deltaf);
	    if (!r) usage();
	    delta = (int) (0.1+deltaf*100);
	    break;
	  default: usage();
	  } 
   }

   if (ParamFile[0])
     read_parameter_file(ParamFile);
   
   if (ns_bases[0]) {
      nonstandards = space(33);
      c=ns_bases;
      i=sym=0;
      if (*c=='-') {
	 sym=1; c++;
      }
      while (*c) {
	 if (*c!=',') {
	    nonstandards[i++]=*c++;
	    nonstandards[i++]=*c;
	    if ((sym)&&(*c!=*(c-1))) {
	       nonstandards[i++]=*c;
	       nonstandards[i++]=*(c-1);
	    }
	 }
	 c++;
      }
   }
   istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
   if ((fold_constrained)&&(istty)) {
      printf("Input constraints using the following notation:\n");
      /* printf("| : paired with another base\n"); */
      printf(". : no constraint at all\n");
      printf("x : base must not pair\n");
      /* printf("< : base i is paired with a base j<i\n"); */
      /* printf("> : base i is paired with a base j>i\n"); */
      /* printf("matching brackets ( ): base i pairs base j\n"); */
   } 
	
   do {				/* main loop: continue until end of file */
      if (istty) {
	 printf("\nInput string (upper or lower case); @ to quit\n");
	 printf("%s\n", scale);
      }
      fname[0]='\0';
      if ((line = get_line(stdin))==NULL) break;

      /* skip comment lines and get filenames */
      while ((*line=='*')||(*line=='\0')||(*line=='>')) {
	if (*line=='>')
	  sscanf(line, ">%20s", fname);
	free(line);
	if ((line = get_line(stdin))==NULL) line = "@";
      } 

      if (strcmp(line, "@") == 0) break;

      sequence = (char *) space(strlen(line)+1);
      sscanf(line,"%s",sequence);
      free(line);
      length = strlen(sequence);

      if (fold_constrained) 
	 cstruc = get_line(stdin);
      
      structure = (char *) space(length+1);
      
      for (l = 0; l < length; l++) sequence[l] = toupper(sequence[l]);
      if (istty)
	printf("length = %d\n", length);

      initialize_fold(length);
      if (fold_constrained) {
	strncpy(structure, cstruc, length+1);
	for (i=0; i<length; i++)
	  if ((structure[i]!='.')&&(structure[i]!='x'))
	    nrerror("only constraints of type 'x' allowed");
      }
      min_en = fold(sequence, structure);
      /* first lines of output (suitable  for sort +1n) */
      if (fname[0] != '\0')
	printf("> %s [%6.2f to %6.2f]\n", fname, min_en, min_en+delta/100.);
      printf("%s %6d %6d\n", sequence, (int) (-0.1+100*min_en), delta); 
       
      subopt(delta);
      
      fflush(stdout);

      free_arrays();
      if (fold_constrained)
	free(cstruc);
      free(sequence);
      free(structure); 
   } while (1);
   return 0;
}

PRIVATE void usage(void)
{
   nrerror("usage: "
	   "RNAsubopt [-e range] [-s] [-lodos] [-logML]\n"
	   "          [-C] [-T temp] [-4] [-d[2]] [-noGU] [-noCloseGU]\n" 
	   "          [-noLP] [-P paramfile] [-nsp pairs]");
}
